RestaurantReservationTest

kikimorella opened the pull request:
add: RestaurantReservationTest

kikimorella committed:
fix: remove all methods and locators that we don't use

kikimorella committed:
add: RestaurantReservationTest

kikimorella commented:
i can't find the commit: https://github.com/violettka/selenide-demo/commit/c52b6f7fe23cc4dcb340e28621e6b0e894f7e0fe

kikimorella commented:
I think the problem is that the new branch is named the same as the remote one. 
what to do? 
create a new branch, add the necessary Java classes to it and push again?

violettka commented on src/main/java/selenide/pages/Homepage.java:
```diff
@@ -15,6 +16,14 @@ public class Homepage extends Page {
+    private static final By profileFiels = By.xpath("//*[@data-qa='header-navigation-profile']");
+
+    private static By restaurantField = By.xpath("//input[@name='dining'])[2]");
+    private static By destinationField = By.xpath("//input[@name='destination'])[2]");
```
у тебя эти локаторы срабатывают? Выглядят, как неработающие

kikimorella replied:
исправила, там потерялась скобка: (//input[@name='dining'])[2]

violettka commented on src/main/java/selenide/pages/Homepage.java:
```diff
@@ -30,4 +39,41 @@ public LoginPage clickOnLoginBtn() {
+    public void fillInTheRestaurant() {
+        homepage.fillDestinationField(RESTAURANT_CITY);
+        homepage.fillRestaurantField(RESTAURANT_NAME);
+        filterPage = homepage.clickOnFindBtn();
```
этот метод должен возвращать фильтер страницу тогда, а не пустоту. Плюс, я бы поменяла название метода. Потому что он делает не только заполнение полей, но и поиск

kikimorella replied:
Готово: fillInTheRestaurantAndClickOnFindBtn()

violettka commented on src/test/java/selenide/RestaurantReservationTest.java:
```diff
@@ -0,0 +1,36 @@
+        loginPage.fillInValidCred();
+        loginPage.clickOnLoginBtnHP();
+        homepage.fillInTheRestaurant();
+        restaurantPage = filterPage.clickOnRestaurant();
```
тут несогласование. ты была на хоумпейдж, а переход на фильтер пейдж нигде не отображен

kikimorella replied:
согласна, строка 29: filterPage = homepage.fillInTheRestaurant();

violettka commented on src/main/java/selenide/pages/FilterPage.java:
```diff
@@ -0,0 +1,18 @@
+
+    // Methods
+    public RestaurantPage clickOnRestaurant() {
+        $(wartesaal).click();
```
та же ошибка, что и у Натальи, - привязка к конкретному ресторану. Я бы советовала привязываться к номеру карточки ресторана. Плюс, если этот метод конкретно создан для клика по этому ресторану, то его название слишком общее. Подумайте о ребятах, которые будут писать тесты для других ресторанов, нужно сделать так, чтобы этот метод был как можно более переиспользуемый.

kikimorella replied:
Теперь привязка к любой пиццерии

violettka commented on src/main/java/selenide/pages/RestaurantPage.java:
```diff
@@ -0,0 +1,40 @@
+    public void reserveNow(String timeFrom, String timeTo) {
+        restaurantPage.dayChoose();
+        restaurantPage.chooseTimeslot(timeFrom, timeTo);
+        reservationPage = restaurantPage.clickOnReserveBtn();
```
та же ошибка с неправильной обработкой возврата новой страницы

kikimorella replied:
исправлено:
public ReservationPage reserveNow(String timeFrom, String timeTo) {
        dayChoose();
        chooseTimeslot(timeFrom, timeTo);
        clickOnReserveBtn();
        return page(ReservationPage.class);
    }

violettka commented on src/main/java/selenide/pages/Page.java:
```diff
@@ -9,9 +9,26 @@
+    RestaurantPage restaurantPage;
+    ReservationPage reservationPage;
+    CheckoutSummaryPage checkoutSummaryPage;
+    UserReservationsPage userReservationsPage;
```
Эти переменные не должны быть здесь. Мы их должны объявлять только в тестовых классах

violettka commented on src/main/java/selenide/pages/UserReservationsPage.java:
```diff
@@ -0,0 +1,22 @@
+
+    // methods
+    public void restaurantIsDisplayed() {
+        $(wartesaal).isDisplayed();
```
я бы использовала shouldBe(displayed)

kikimorella replied:
потому что: **The isDisplayed method in Selenium** verifies if a certain element is present and displayed?

возможно, лучше: should(exist)

kikimorella committed:
fix: RestaurantReservationTest making changes according to the comments

kikimorella committed:
fix: moving header locators in class Page

kikimorella committed:
fix: 1. FilterPage - replaced a specific restaurant with any first
2. ReservationPage renamed to CompleteReservationPage
3. UserReservationsPage - any restaurant should be in the field Upcoming Reservations and the Discover Restaurants button should be missing
4. CheckoutSummaryPage - we use restaurantName to confirm that the new restaurant is reserved

kikimorella commented on src/main/java/selenide/pages/UserReservationsPage.java:
```diff
@@ -8,16 +8,15 @@
-    private static final By wartesaal = By.xpath("//a[@href='/place/wartesaal-11828']/h3");
+    private static final By upcomingReservations = By.xpath("//*[@data-qa='cards-upcoming-reservations']");
+    private static final By discoverRestaurantsBtn = By.xpath("//*[@data-qa='reservations-cta-reserve']");
 
```
Вопрос прежний, как в тесте оформлять preconditions. Тут нужно, чтоб не было активных/бедующих резерваций, тогда будет присутствовать кнопка discoverRestaurantsBtn. Можно будет отдельным методом перед тестом проверять её наличие, а потом anyRestaurantIsDisplayed соответственно её отсутствие.

kikimorella commented on src/main/java/selenide/pages/CheckoutSummaryPage.java:
```diff
@@ -1,20 +1,27 @@
+        specialName = $(restaurantName).getText();
         $(loginIcon).click();
         $(reservations).click();
+        $(upcomingReservations).shouldHave(Condition.text(specialName));
```
муж сказал, что при параллельном запуске тестов возникнет RaceCondition

Смущает, что upcomingReservations находится на другой странице UserReservationsPage. Как провести проверку по другому, я не знаю.

kikimorella commented on src/test/java/selenide/RestaurantReservationTest.java:
```diff
@@ -0,0 +1,36 @@
+        homepage.accCookies();
+        loginPage = homepage.clickOnLoginBtn();
+        loginPage.fillInValidCred();
+        loginPage.clickOnLoginBtnHP();
```
мы объявляем переменную 1 раз, уже объявлено: homepage = open(BASE_URL, Homepage.class);

kikimorella commented on src/main/java/selenide/pages/CheckoutSummaryPage.java:
```diff
@@ -0,0 +1,27 @@
+
+    // locators
+    private static final By loginIcon = By.xpath("//*[@data-qa='header-navigation-button");
+    private static final By reservations = By.xpath("//*[@data-qa='header-navigation-button");
```
изменила: private static final By reservations = By.xpath("//*[@data-qa='header-navigation-reservations']");

kikimorella committed:
fix: reservations locator

kikimorella committed:
fix: we use git rebase

violettka commented on src/main/java/selenide/pages/Homepage.java:
```diff
@@ -37,7 +38,8 @@ public void fillRestaurantField(String what) {
 
     public void fillDestinationField(String where) {
-        $(destinationField).sendKeys(where);
+        $(destinationField).sendKeys(where); // нужно ожидание, т.к. не успевает меняться город
```
Мы говорили о том, чтобы добавить шаги выбора города из дропдауна

kikimorella committed:
Add: test, which checks unsuccessful entry

kikimorella committed:
Fix: error, in the test you can not set the time range when choosing a restaurant
